### DISCUSSION TITLE
Motion correction parameter to maintain the same image size

### DISCUSSION THREAD

**Discussion by kfouke1** [flatironinstitute/CaImAn#1244] [Q&A] [answered]
Hi all,

I was wondering if there is any motion correction parameter that would make sure that the motion corrected image maintains the same pixel size as the original image? i.e. my original image is 512 x 512 px, but then corrected it turns into 504 x 504 px. I tried "border_nan": True, but it still does not keep the same size.

Thanks in advance!
Kaitlyn

**Comment by EricThomson** [comment]
The `border_nan` param determines how to handle the nans that result when the movie is shifted at the edges. If you set it to `True` it leaves the `NaN`s in there, which may lead to cropping later. I would set that to `copy`, which fills in the `NaN`s with real values. If you set it to `copy` and still are getting cropping then that would definitely be a concern.

**Reply by kfouke1** [reply]
Hi Eric,

Thanks for your response! Hmm, I do normally set it to 'copy' but it will still crop the image (508 x 508). The caiman version that I like to use is 1.9.8, so I'm not sure if that would make a difference. Here are the caiman ops that I set - would any of these have an effect?

caiman_ops = {
            "max_shifts": (3, 3),
            "strides": (25, 25),
            "overlaps": (15, 15),
            "num_frames_split": 150,
            "max_deviation_rigid": 3,
            "pw_rigid": False,
            "shifts_opencv": True,
            "border_nan": "copy",
            "downsample_ratio": 0.2,
        }

Thanks!
Kaitlyn

**Reply by EricThomson** [reply]
That is surprising to me, but there are often things that come up unexpectedly in caiman. :smile:  So you are running CNMF, not CNMFE?  For instance, whenever I've run motion correction and looked at original movie shape, and then later looked at the images shape of the memory mapped file, they have always been the same. Let me find a representative set of params I use and I'll add them in here in a minute. 

Oh wait what are you giving for `border_to_0`? Is that `4` by any chance? 

We have something like this in the cnmf demo:
```
border_to_0 = 0 if mot_correct.border_nan == 'copy' else mot_correct.border_to_0 # trim border against NaNs
mc_memmapped_fname = cm.save_memmap(mot_correct.mmap_file, 
                                        base_name='memmap_', 
                                        order='C',
                                        border_to_0=border_to_0,  # exclude borders, if that was done
                                        dview=cluster)
```

The edges are sort of trash anyway, so we sometimes trim them with this, but you may have previously had it not to `copy` and then not turned it back or something. 


Edit: I just tested with `border_to_0` set to 10 and it didn't crop the movie it just set it to 0, blacking it out at those borders. So that's not it. I have no idea what actually crops it in our code base except certain spatial cropping params. I'd make sure you aren't accidentally setting them somewhere. Feel free to share your script I could take a look, but that's the only thing that explicitly changes dims that I'm aware of.

**Reply by kfouke1** [reply, marked as answer]
With this data, I have actually been running Suite2p for source extraction. These motion correction parameters are actually inherited by Matt, which works on the data that we collect, but for my photostimulation work I do need the full pixel size for analysis. 

This is the code that I run for motion correction:

```
def run_movement_correction(
    base_fish,
    caiman_ops=None,
    keep_mmaps=False,
    force=False,
):
    import caiman as cm
    from tifffile import imsave   

if "rotated_image" in base_fish.data_paths.keys():
        original_image_path = base_fish.data_paths["rotated_image"]
    else:
        original_image_path = base_fish.data_paths["image"]

if not caiman_ops:
        caiman_ops = {
            "max_shifts": (3, 3),
            "strides": (25, 25),
            "overlaps": (15, 15),
            "num_frames_split": 150,
            "max_deviation_rigid": 3,
            "pw_rigid": False,
            "shifts_opencv": True,
            "border_nan": "copy",
            "downsample_ratio": 0.2,
        }
    c, dview, n_processes = cm.cluster.setup_cluster(
        backend="local", n_processes=12, single_thread=False
    )
    mc = cm.motion_correction.MotionCorrect(
        [original_image_path.as_posix()],
        dview=dview,
        max_shifts=caiman_ops["max_shifts"],
        strides=caiman_ops["strides"],
        overlaps=caiman_ops["overlaps"],
        max_deviation_rigid=caiman_ops["max_deviation_rigid"],
        shifts_opencv=caiman_ops["shifts_opencv"],
        nonneg_movie=True,
        border_nan=caiman_ops["border_nan"],
        is3D=False,
    )
    mc.motion_correct(save_movie=True)
    bord_px_rig = np.ceil(np.max(mc.shifts_rig)).astype(np.int)
    mc.pw_rigid = True  # turn the flag to True for pw-rigid motion correction
    mc.template = (mc.mmap_file )  # use the template obtained before to save in computation (optional)
    mc.motion_correct(save_movie=True, template=mc.total_template_rig)
    m_els = cm.load(mc.fname_tot_els)
    output = m_els[
        :,
        2 * bord_px_rig : -2 * bord_px_rig,
        2 * bord_px_rig : -2 * bord_px_rig,
    ]
    if not keep_mmaps:
        with os.scandir(original_image_path.parents[0]) as entries:
            for entry in entries:
                if entry.is_file():
                    if entry.name.endswith(".mmap"):
                        os.remove(entry)
    dview.terminate()
    cm.stop_server()
```

Now looking at this, I wonder if the output is cropped here automatically from this line:  output = m_els[:, 2 * bord_px_rig : -2 * bord_px_rig, 2 * bord_px_rig : -2 * bord_px_rig]....but maybe you can read this better than I can!

Thanks so much!

**Reply by EricThomson** [reply]
Yes that is likely doing it. You can check `m_els.shape` and `output.shape` and they are likely different. It looks like someone added in slicing there for some reason.

**Reply by kfouke1** [reply]
Woohoo yes that was it! Thanks for your help - even since it was not a motion correction issue in the first place. Sometimes sharing code helps!
